File: rotate.py
#Approach 1- Rotate groups of 4 cells 
def rotate(matrix):   # TC O(M) // SC O(1)   M--> number of cells in the matrix 
	n = len(matrix[0])
	for i in range(n // 2 + n % 2):
	    for j in range(n // 2):
	        tmp = matrix[n - 1 - j][i]
	        matrix[n - 1 - j][i] = matrix[n - 1 - i][n - j - 1]
	        matrix[n - 1 - i][n - j - 1] = matrix[j][n - 1 -i]
	        matrix[j][n - 1 - i] = matrix[i][j]
	        matrix[i][j] = tmp
	return matrix

'''
Approach 2- Reverse on Diagonal and then reverse left to right 
First Transpose
Then reverse 

''' 
def rotate1(matrix):  # TC O(M) // SC O(1) M--> number of cells in the matrix 
    transpose(matrix)
    reflect(matrix) # reverse 
    return matrix 

def transpose(matrix):
    n = len(matrix)
    for i in range(n):
        for j in range(i + 1, n):
            matrix[j][i], matrix[i][j] = matrix[i][j], matrix[j][i]

def reflect(matrix):
    n = len(matrix)
    for i in range(n):
        for j in range(n // 2):
            matrix[i][j], matrix[i][-j - 1] = matrix[i][-j - 1], matrix[i][j]



print(rotate([[1,2,3],[4,5,6],[7,8,9]]))
print(rotate1([[1,2,3],[4,5,6],[7,8,9]]))

File: rotate.java
import java.util.*;

//Approach 1- Transpose and reflect 
class Solution { // TC O(m)  SC O(1) m--> no. of cells 
    public String rotate(int[][] matrix) {
        transpose(matrix);
        reflect(matrix);
        return Arrays.deepToString(matrix);
    }
    
    public void transpose(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                int tmp = matrix[j][i];
                matrix[j][i] = matrix[i][j];
                matrix[i][j] = tmp;
            }
        }
    }
    
    public void reflect(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n / 2; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[i][n - j - 1];
                matrix[i][n - j - 1] = tmp;
            }
        }
    }
}

//Approach 2- Rotate groups of 4 cells
class Solution1{ // TC O(m)  SC O(1) m--> no. of cells 
    public String rotate(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < (n + 1) / 2; i ++) {
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[n - 1 - j][i];
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - j - 1];
                matrix[n - 1 - i][n - j - 1] = matrix[j][n - 1 -i];
                matrix[j][n - 1 - i] = matrix[i][j];
                matrix[i][j] = temp;
            }
        }
        return Arrays.deepToString(matrix);
    }
}

class rotate{
    public static void main(String[] args) {
        int [][] matrix={{1,2,3},{4,5,6},{7,8,9}};
        Solution s= new Solution();
        Solution1 s1= new Solution1();
        System.out.println(s.rotate(matrix));
        System.out.println(s1.rotate(matrix));
    }
}